File: khalti-android/src/main/java/com/khalti/checkOut/CheckOutContract.java
package com.khalti.checkOut;

import java.util.List;

import com.khalti.checkOut.helper.MerchantPreferencePojo;
import com.khalti.checkOut.helper.PaymentPreference;

import rx.Observable;

interface CheckOutContract {
    interface View {

        void toggleIndented(boolean show);

        void showIndentedNetworkError();

        void showIndentedError(String error);

        Observable<Void> setTryAgainClickListener();

        void setupViewPager(List<PaymentPreference> types);

        void setUpTabLayout(List<PaymentPreference> types);

        void setTabListener();

        void toggleTab(int position, boolean selected);

        void setStatusBarColor();

        void dismissAllDialogs();

        boolean hasNetwork();

        void closeCheckOut();

        void setPresenter(Presenter presenter);
    }

    interface Presenter {

        void onCreate();

        void onDestroy();

        void onTabSelected(int position, boolean selected);

        void fetchPreference(String key);
    }

    interface Model {

        Observable<MerchantPreferencePojo> fetchPreference(String key);

        void unSubscribe();
    }
}

File: khalti-android/src/main/java/com/khalti/checkout/CheckOutContract.kt
package com.khalti.checkout

import com.khalti.base.LifeCycle
import com.khalti.checkout.helper.Config
import com.khalti.checkout.helper.PaymentPreference
import com.khalti.signal.Signal

interface CheckOutContract {
    interface View {

        fun toggleToolbarShadow(show: Boolean)

        fun toggleTitle(show: Boolean)

        fun toggleSearch(paymentType: String, show: Boolean)

        fun toggleLoading(show: Boolean)

        fun toggleTestBanner(show: Boolean)

        fun setupViewPager(types: List<PaymentPreference>)

        fun setUpTabLayout(types: List<PaymentPreference>): Signal<Map<String, Any>>

        fun setPageScrollListener(currentPage: Int): Signal<Int>

        fun setStatusBarColor()

        fun dismissAllDialogs()

        fun closeCheckOut()

        fun convertDpToPx(dp: Int): Int

        fun getSearchViewMapInitSignal(): Signal<Any>

        fun getAppPackageName() : String

        fun setPresenter(presenter: Presenter)
    }

    interface Presenter : LifeCycle {

        fun onGetPreferenceList(config: Config): List<PaymentPreference>

        fun onTabSelected(preferences: List<PaymentPreference>, it: Map<String, Any>)

        fun onBackPressed()
    }
}
